acceptancetests/deploy_stack.py

```diff
@@ -570,6 +571,68 @@ def tear_down(self, has_controller):
             self.tear_down_client.kill_controller(check=True)
 
 
+class ExistingController:
+    """A Controller strategy where the controller is already present.
+
+    Intended for use with BootstrapManager and
+    version_client.client_for_existing(). The ivar env in this strategy is a
+    bit of a hack to allow jujupy to set up and track the correct model name
+    to test under.
+
+    :ivar client: Client object
+    :ivar tear_down_client: Client object to tear down at the end of testing
+    """
+
+    def __init__(self, client):
+        self.client = client
+        self.tear_down_client = client
+
+    def create_initial_model(self):
+        """Create the initial model for use in testing.
+
+        Since we set client.env.environment to our desired model name jujupy
+        picks that up to name the new model.
+        """
+
+        self.client.add_model(self.client.env)
+        logging.info('Added model {} to existing controller'.format(
+            self.client.env.environment))
+
+    def prepare(self, controller_id):
+        """Prepare client for use by pointing it at the selected controller.
+
+        This is a bit of a hack to allow for multiple controllers in the same
+        environment while testing. When the client object is intiailly made out
+        of the existing environment it picks up the current controller and
+        sets the env.controller name to that ID. Resetting the name to occurred
+        desired ID simply forces jujupy to pass that ID as the first part of
+        the -m <controller>:<model> flag for commands.
+
+        :param controller_id: ID of the controller in use for testing, passed
+        in with the --existing flag
+        """
+        self.client.env.controller.name = controller_id
+
+    def get_hosts(self):
+        """Provide the controller host."""
+        host = get_machine_dns_name(
+            self.client.get_controller_client(), '0')
+        if host is None:
+            raise ValueError('Could not get machine 0 host')
+        return {'0': host}
+
+    def tear_down(self, _):
+        """Destroy the current model and switch back to a known default model.
+
+        The client.switch allows another client_from_existing client object to
+        be created after this test completes, in the event that tests are to be
+        run in serial.
+        """
+        self.client.destroy_model()
+        self.client.switch(
+            'controller', controller=self.client.env.controller.name)
```

Comment:
Thinking about this, setting the current model while other tests are running using the same JUJU_DATA (i.e. concurrent tests) might muck up the tests if the current model the are using is swapped out from underneath them.

That being said I'm pretty sure the amount of commands (let alone tests) that rely on the model being select should be approaching zero (you've identified a place where it's expected) so it shouldn't impact anything.

The 'problem' I have is that it's being set when it probably doesn't need to be.
  - If your test is using a shared JUJU_DATA it's either to use the --existing functionality or just sharing the resource in which case you can't guarantee the current model isn't going to change (adding a model for instance changes this).
  - If you're not using --existing you're probably not using a shared JUJU_DATA anyway (and if you are see above about there being no promise the current model won't change from out from underneath you.
  - If you are using --existing and client_for_existing gets called (i.e. in the one test that uses it) the model and controller name can be provided so there is no need to have a selected model (esp if you take note of the comment below stating you don't need to get current if you're provided model and controller name).

All that to say, I'm pretty sure you can just remove the switch as it won't impact anything.